README.md
#### What is this?
Syncthing ([core](https://github.com/syncthing/syncthing)) uses a rescan interval to detect changes in folders. This application (syncthing-inotify) uses OS primitives to detect changes as soon as they happen. Therefore, if you save a file, syncthing-inotify will know about it and pass this information to Syncthing such that near real-time synchronisation can be achieved.

#### Where do I get it?
Syncthing-inotify binaries are released on [github](https://github.com/syncthing/syncthing-inotify/releases/latest)

#### How do I run it?
Syncthing-inotify will automatically read the Syncthing config if it is found in the standard path.
  * Run and hide on unix in a screen
```
screen -S inotify -dm ./syncthing-inotify
```
  * Run and hide on windows in .vbs script
```
CreateObject("Wscript.Shell").Run "syncthing-inotify.exe, 0, True
```
  * Run and hide on windows using API key in .vbs script
```
CreateObject("Wscript.Shell").Run "syncthing-inotify.exe -api=""...""", 0, True
```
  * Install as a service, see the etc/ folder

#### I'm confused
  * Try [Syncthing-GTK](https://github.com/syncthing/syncthing-gtk)
  * Read the commandline options: ```./syncthing-inotify -help```. Settings, such as an API key, need to be manually provided if you use a custom home for Syncthing.

#### Building syncthing-inotify
```
# To clone
mkdir -p src/github.com/syncthing
git clone https://github.com/syncthing/syncthing-inotify.git src/github.com/syncthing/syncthing-inotify
# Following commands are needed every time you want to build (unless you use Golang's specific folder structure: C:\src  or ~/src/)
export GOPATH=$(pwd)
cd src/github.com/syncthing/syncthing-inotify
go get
go build
# To tag the build with a version number, use
go build -ldflags "-w -X main.Version `git describe --abbrev=0 --tags`"
# See cross-compile.rb for further compilation details
# - Please prefix the build with GO386=387 when building for i386
# - Please prefix the build with GOARM=5 when building for arm
# - Depends on darwin_amd64/notify.a for OSX builds
```


#### Troubleshooting for folders with many files on Linux
* Linux limits the amount of inotify watchers (typically to [8192](http://stackoverflow.com/a/20355253)). Therefore, if you wish to sync many files and folders, you'll need to increase the upper limit:

  Permanently fix `Too many open files` for Linux: ```sudo sh -c 'echo fs.inotify.max_user_watches=204800\n >> /etc/sysctl.conf'```

  Fix `Too many open files` for Linux until next reboot: ```sudo sh -c 'echo 204800 > /proc/sys/fs/inotify/max_user_watches'``` (should be applied before launching syncthing-inotify)


debian/changelog
syncthing-inotify ({{version}}); urgency=medium

  * Packaging of {{version}}.

   -- {{mntner}}  {{date}}


syncwatcher_test.go
// syncwatcher_test.go
package main

import (
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"testing"
	"time"
)

var (
	slash         = string(os.PathSeparator)
	testDirectory = "test"
)

func initTestDir() {
	os.RemoveAll(testDirectory)
	os.MkdirAll(testDirectory, 0755)
}

func createTestPaths(t *testing.T, dir string, fs ...string) []string {
	rs := make([]string, len(fs))
	for i, f := range fs {
		rs[i] = createTestPath(t, dir, f)
	}
	return rs
}

func createTestPath(t *testing.T, dir, f string) string {
	if strings.HasSuffix(f, slash) {
		err := os.MkdirAll(dir+slash+f, 0755)
		if err != nil && !os.IsExist(err) {
			t.Error("Failed to create test directory", err)
		}
		return strings.TrimSuffix(f, slash)
	} else {
		err := os.MkdirAll(filepath.Dir(dir+slash+f), 0755)
		if err != nil && !os.IsExist(err) {
			t.Error("Failed to create test directory", err)
		}
	}
	h, err := os.Create(dir + slash + f)
	if err != nil {
		t.Error("Failed to create test file", err)
	}
	h.Close()
	return f
}

func TestDebouncedFileWatch(t *testing.T) {
	// Log file change
	testOK := false
	testRepo := "test1"
	testFile := "a" + slash + "file1"
	testFiles := createTestPaths(t,
		testFile)
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 10
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 1 || sub[0] != testFile {
			t.Error("Invalid result for file change: "+repo, sub)
		}
		testOK = true
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for i := range testFiles {
		fsChan <- testDirectory + slash + testFiles[i]
	}
	time.Sleep(testDebounceTimeout * 10)
	if !testOK {
		t.Error("Callback not triggered")
	}
}

func TestDebouncedDirectoryWatch(t *testing.T) {
	// Log directory change
	testOK := false
	testRepo := "test1"
	testFile := createTestPath(t, testDirectory, "a"+slash)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 10
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 1 || sub[0] != testFile {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		testOK = true
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	fsChan <- testDirectory + slash + testFile
	time.Sleep(testDebounceTimeout * 10)
	if !testOK {
		t.Error("Callback not triggered")
	}
}

func TestDebouncedParentDirectoryWatch(t *testing.T) {
	// Convert a/file1.txt a/file2 a/file3.ogg to a
	testOK := false
	testRepo := "test1"
	testChangeDir := "a" + slash
	testFiles := createTestPaths(t,
		testChangeDir+"file1.txt",
		testChangeDir+"file2",
		testChangeDir+"file3.ogg")
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 2
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 1 || sub[0] != "a" {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		testOK = true
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for i := range testFiles {
		fsChan <- testDirectory + slash + testFiles[i]
	}
	time.Sleep(testDebounceTimeout * 10)
	if !testOK {
		t.Error("Callback not triggered")
	}
}

func TestDebouncedParentDirectoryWatch2(t *testing.T) {
	// Convert a a/file1.txt a/file2 b a/file3.ogg to a b
	testOK := 0
	testRepo := "test1"
	testChangeDir1 := "a" + slash
	testChangeDir2 := "b" + slash
	testFiles := createTestPaths(t,
		testChangeDir1,
		testChangeDir1+"file1.txt",
		testChangeDir1+"file2",
		testChangeDir2,
		testChangeDir1+"file3.ogg")
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 10
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 2 || sub[0] != "a" {
			t.Error("Invalid result for directory change 1: "+repo, sub)
		}
		if repo != testRepo || sub[1] != "b" {
			t.Error("Invalid result for directory change 2: "+repo, sub)
		}
		testOK = len(sub)
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for i := range testFiles {
		fsChan <- testDirectory + slash + testFiles[i]
	}
	time.Sleep(testDebounceTimeout * 10)
	if testOK != 2 {
		t.Error("Callback not correctly triggered")
	}
}

func TestDebouncedParentDirectoryWatch3(t *testing.T) {
	// Don't convert a/b/file1.txt a/c/file2 a/d/file3.ogg
	testOK := 0
	testRepo := "test1"
	testFiles := createTestPaths(t,
		"a"+slash+"b"+slash+"file1.txt",
		"a"+slash+"c"+slash+"file2",
		"a"+slash+"d"+slash+"file3.ogg")
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 3
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		for i, s := range sub {
			if repo != testRepo || s != testFiles[i] {
				t.Error("Invalid result for directory change " + strconv.Itoa(testOK) + ": " + repo + " " + s)
			}
		}
		testOK = len(sub)
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for i := range testFiles {
		fsChan <- testDirectory + slash + testFiles[i]
	}
	time.Sleep(testDebounceTimeout * 10)
	if testOK != 3 {
		t.Error("Callback not correctly triggered")
	}
}

func TestDebouncedParentDirectoryWatch4(t *testing.T) {
	// Convert a/e a/b/d a/b/file1.txt a/b/file2 a/b/file3.ogg a/b/c/file4 to a/b a/e
	testOK := 0
	testRepo := "test1"
	testFiles := createTestPaths(t,
		"a"+slash+"e",
		"a"+slash+"b"+slash+"d",
		"a"+slash+"b"+slash+"file1.txt",
		"a"+slash+"b"+slash+"file2",
		"a"+slash+"b"+slash+"file3.ogg",
		"a"+slash+"b"+slash+"c"+slash+"file4")
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 3
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 2 || sub[0] != "a"+slash+"b" {
			t.Error("Invalid result for directory change "+strconv.Itoa(testOK)+": "+repo, sub)
		}
		if repo != testRepo || sub[1] != "a"+slash+"e" {
			t.Error("Invalid result for directory change "+strconv.Itoa(testOK)+": "+repo, sub)
		}
		testOK = len(sub)
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for i := range testFiles {
		fsChan <- testDirectory + slash + testFiles[i]
	}
	time.Sleep(testDebounceTimeout * 10)
	if testOK != 2 {
		t.Error("Callback not correctly triggered")
	}
}

func TestDebouncedParentDirectoryWatch5(t *testing.T) {
	// Convert a/b a/c file1 file2 file3 to _ (main folder)
	testOK := false
	testRepo := "test1"
	testFiles := createTestPaths(t,
		"a"+slash+"b"+slash,
		"a"+slash+"c"+slash,
		"file1",
		"file2",
		"file3")
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 3
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 1 || sub[0] != "" {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		testOK = true
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for i := range testFiles {
		fsChan <- testDirectory + slash + testFiles[i]
	}
	time.Sleep(testDebounceTimeout * 10)
	if !testOK {
		t.Error("Callback not correctly triggered")
	}
}

func TestDebouncedParentDirectoryWatch6(t *testing.T) {
	// Convert a/b/c a/b/c/f1 a/b/c/f2 a/b/c/f3 to a/b/c
	testOK := 0
	testRepo := "test1"
	testChangeDir := "a" + slash + "b" + slash + "c" + slash
	testFiles := createTestPaths(t,
		testChangeDir,
		testChangeDir+"file1.txt",
		testChangeDir+"file2",
		testChangeDir+"file3.ogg")
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 10
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 1 || sub[0] != strings.TrimSuffix(testChangeDir, slash) {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		testOK += 1
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for i := range testFiles {
		fsChan <- testDirectory + slash + testFiles[i]
	}
	time.Sleep(testDebounceTimeout * 10)
	if testOK != 1 {
		t.Error("Callback not correctly triggered")
	}
}

func TestDebouncedParentDirectoryRemovedWatch(t *testing.T) {
	// Convert a a/b a/b/test.txt into a
	testOK := 0
	testRepo := "test1"
	testFiles := createTestPaths(t,
		"a"+slash,
		"a"+slash+"b"+slash,
		"a"+slash+"b"+slash+"file1.txt")
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 10
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 1 || sub[0] != "a" {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		testOK += 1
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for i := range testFiles {
		fsChan <- testDirectory + slash + testFiles[i]
	}
	time.Sleep(testDebounceTimeout * 10)
	if testOK != 1 {
		t.Error("Callback not correctly triggered")
	}
}

func TestSTEvents(t *testing.T) {
	// Ignore notifications if ST created them
	testOK := true
	testRepo := "test1"
	testFiles := createTestPaths(t,
		"file1",
		"file2",
		"file3")
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 10
	stChan := make(chan STEvent, 10)
	fsChan := make(chan string, 10)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 0 {
			t.Error("Invalid result for directory change: " + repo)
		}
		testOK = false
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	stChan <- STEvent{Path: ""}
	for i := range testFiles {
		stChan <- STEvent{Path: testDirectory + slash + testFiles[i], Finished: false}
		fsChan <- testDirectory + testFiles[i]
		stChan <- STEvent{Path: testDirectory + slash + testFiles[i], Finished: true}
	}
	time.Sleep(testDebounceTimeout * 50)
	if !testOK {
		t.Error("Callback not correctly triggered")
	}
}

func TestRealFolderPath(t *testing.T) {
	nrFiles := 50
	testOK := false
	testRepo := "test1"
	testFiles := make([]string, nrFiles)
	oldDir := testDirectory
	defer func() { testDirectory = oldDir }()
	root, err := os.Getwd()
	if err != nil {
		t.Error("Failed to retrieve pwd", err)
	}
	testDirectory = root + "/tmp/testSyncthingInotify"
	if err := os.Mkdir(testDirectory, 0777); err != nil {
		t.Log("Creating " + testDirectory + " failed, skipping test")
		t.Skip()
	}
	for i := 0; i < nrFiles; i++ {
		testFiles[i] = createTestPath(t, testDirectory, "a"+slash+strconv.Itoa(i))
	}
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := nrFiles + 1
	stop := make(chan int, 1)
	stChan := make(chan STEvent, nrFiles)
	fsChan := make(chan string, nrFiles)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 50 || sub[0] != "a/0" {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		if testOK {
			t.Error("Callback triggered multiple times")
		}
		testOK = true
		close(stop)
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for _, testFile := range testFiles {
		fsChan <- testDirectory + slash + testFile
	}
	<-stop
	time.Sleep(250 * time.Millisecond)
	if !testOK {
		t.Error("Callback not triggered")
	}
}

func TestRealSymlinkFolderPath(t *testing.T) {
	nrFiles := 50
	testOK := false
	testRepo := "test1"
	testFiles := make([]string, nrFiles)
	oldDir := testDirectory
	defer func() { testDirectory = oldDir }()
	root, err := os.Getwd()
	if err != nil {
		t.Error("Failed to retrieve pwd", err)
	}
	testDirectory = root + "/tmp/testSyncthingInotify"
	testLink := root + "/tmp/testSyncthingInotifyLn"
	if err := os.Mkdir(testDirectory, 0777); err != nil {
		t.Log("Creating " + testDirectory + " failed, skipping test")
		t.Skip()
	} else {
		defer os.RemoveAll(testDirectory)
	}
	if err := os.Symlink(testDirectory, testLink); err != nil {
		t.Log("Creating symlink " + testLink + " failed, skipping test")
		t.Skip()
	} else {
		defer os.RemoveAll(testLink)
	}
	for i := 0; i < nrFiles; i++ {
		testFiles[i] = createTestPath(t, testLink, "a"+slash+strconv.Itoa(i))
	}
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := nrFiles + 1
	stop := make(chan int, 1)
	stChan := make(chan STEvent, nrFiles)
	fsChan := make(chan string, nrFiles)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 50 || sub[0] != "a/0" {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		if testOK {
			t.Error("Callback triggered multiple times")
		}
		testOK = true
		close(stop)
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for _, testFile := range testFiles {
		fsChan <- testLink + slash + testFile
	}
	<-stop
	time.Sleep(250 * time.Millisecond)
	if !testOK {
		t.Error("Callback not triggered")
	}
}
func TestFilesAggregation(t *testing.T) {
	nrFiles := 50
	testOK := false
	testRepo := "test1"
	testFiles := make([]string, nrFiles)
	for i := 0; i < nrFiles; i++ {
		testFiles[i] = createTestPath(t, testDirectory, "a"+slash+strconv.Itoa(i))
	}
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := nrFiles + 1
	stop := make(chan int, 1)
	stChan := make(chan STEvent, nrFiles)
	fsChan := make(chan string, nrFiles)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 50 || sub[0] != "a/0" {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		if testOK {
			t.Error("Callback triggered multiple times")
		}
		testOK = true
		close(stop)
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for _, testFile := range testFiles {
		fsChan <- testDirectory + slash + testFile
	}
	<-stop
	time.Sleep(250 * time.Millisecond)
	if !testOK {
		t.Error("Callback not triggered")
	}
}

func TestManyFilesAggregation(t *testing.T) {
	nrFiles := 5000
	testOK := false
	testRepo := "test1"
	testFiles := make([]string, nrFiles)
	for i := 0; i < nrFiles; i++ {
		testFiles[i] = createTestPath(t, testDirectory, "a"+slash+strconv.Itoa(i))
	}
	defer os.RemoveAll(testDirectory)
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 10
	stop := make(chan int, 1)
	stChan := make(chan STEvent, nrFiles)
	fsChan := make(chan string, nrFiles)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 1 || sub[0] != "" {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		if testOK {
			t.Error("Callback triggered multiple times")
		}
		testOK = true
		close(stop)
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for _, testFile := range testFiles {
		fsChan <- testDirectory + slash + testFile
	}
	<-stop
	time.Sleep(250 * time.Millisecond)
	if !testOK {
		t.Error("Callback not triggered")
	}
}

func TestDeletesAggregation(t *testing.T) {
	nrFiles := 50
	testOK := false
	testRepo := "test1"
	testFiles := make([]string, nrFiles)
	for i := 0; i < nrFiles; i++ {
		testFiles[i] = "a" + slash + strconv.Itoa(i)
	}
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 10
	stop := make(chan int, 1)
	stChan := make(chan STEvent, nrFiles)
	fsChan := make(chan string, nrFiles)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 50 || sub[0] != "a/0" {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		if testOK {
			t.Error("Callback triggered multiple times")
		}
		testOK = true
		close(stop)
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for _, testFile := range testFiles {
		fsChan <- testDirectory + slash + testFile
	}
	<-stop
	time.Sleep(250 * time.Millisecond)
	if !testOK {
		t.Error("Callback not triggered")
	}
}
func TestManyDeletesAggregation(t *testing.T) {
	nrFiles := 5000
	testOK := false
	testRepo := "test1"
	testFiles := make([]string, nrFiles)
	for i := 0; i < nrFiles; i++ {
		testFiles[i] = "a" + slash + strconv.Itoa(i)
	}
	testDebounceTimeout := 20 * time.Millisecond
	testDirVsFiles := 10
	stop := make(chan int, 1)
	stChan := make(chan STEvent, nrFiles)
	fsChan := make(chan string, nrFiles)
	fileChange := func(repo string, sub []string) error {
		if len(sub) == 1 && sub[0] == ".stfolder" {
			return nil
		}
		if repo != testRepo || len(sub) != 1 || sub[0] != "" {
			t.Error("Invalid result for directory change: "+repo, sub)
		}
		if testOK {
			t.Error("Callback triggered multiple times")
		}
		testOK = true
		close(stop)
		return nil
	}
	go accumulateChanges(testDebounceTimeout, testRepo, testDirectory, testDirVsFiles, stChan, fsChan, fileChange)
	for _, testFile := range testFiles {
		fsChan <- testDirectory + slash + testFile
	}
	<-stop
	time.Sleep(250 * time.Millisecond)
	if !testOK {
		t.Error("Callback not triggered")
	}
}


syncwatcher_unix_test.go
// syncwatcher_test.go
//+build !windows

package main

import (
	"encoding/json"
	"regexp"
	"strings"
	"testing"
)

func TestIgnores(t *testing.T) {
	s := `{"patterns":["(?i)^\\.DS_Store$","(?i)^.*\\\\\\.DS_Store$","(?i)^\\.DS_Store\\\\.*$","(?i)^.*\\\\\\.DS_Store\\\\.*$","(?i)^\\.Spotlight-V100$","(?i)^.*\\\\\\.Spotlight-V100$","(?i)^\\.Spotlight-V100\\\\.*$","(?i)^.*\\\\\\.Spotlight-V100\\\\.*$","(?i)^\\.Trashes$","(?i)^.*\\\\\\.Trashes$","(?i)^\\.Trashes\\\\.*$","(?i)^.*\\\\\\.Trashes\\\\.*$","(?i)^~[^\\\\]*$","(?i)^.*\\\\~[^\\\\]*$","(?i)^~[^\\\\]*\\\\.*$","(?i)^.*\\\\~[^\\\\]*\\\\.*$","(?i)^ehthumbs\\.db$","(?i)^.*\\\\ehthumbs\\.db$","(?i)^ehthumbs\\.db\\\\.*$","(?i)^.*\\\\ehthumbs\\.db\\\\.*$","(?i)^desktop\\.ini$","(?i)^.*\\\\desktop\\.ini$","(?i)^desktop\\.ini\\\\.*$","(?i)^.*\\\\desktop\\.ini\\\\.*$","(?i)^Thumbs\\.db$","(?i)^.*\\\\Thumbs\\.db$","(?i)^Thumbs\\.db\\\\.*$","(?i)^.*\\\\Thumbs\\.db\\\\.*$","(?i)^\\._[^\\\\]*$","(?i)^.*\\\\\\._[^\\\\]*$","(?i)^\\._[^\\\\]*\\\\.*$","(?i)^.*\\\\\\._[^\\\\]*\\\\.*$","(?i)^\\.sync$","(?i)^.*\\\\\\.sync$","(?i)^\\.sync\\\\.*$","(?i)^.*\\\\\\.sync\\\\.*$","(?i)^\\.git$","(?i)^.*\\\\\\.git$","(?i)^\\.git\\\\.*$","(?i)^.*\\\\\\.git\\\\.*$","(?i)^\\.Trash-1000$","(?i)^.*\\\\\\.Trash-1000$","(?i)^\\.Trash-1000\\\\.*$","(?i)^.*\\\\\\.Trash-1000\\\\.*$","(?i)^ignored folder$","(?i)^ignored file\\..*$"]}`
	var ignores map[string][]string
	err := json.Unmarshal([]byte(s), &ignores)
	if err != nil {
		t.Fatal(err)
	}
	ignorePaths := ignores["ignore"]
	ignorePatterns := make([]Pattern, len(ignores["patterns"]))
	for i, str := range ignores["patterns"] {
		pattern := strings.TrimPrefix(str, "(?exclude)")
		regexp, err := regexp.Compile(pattern)
		if err != nil {
			t.Fatal(err)
		}
		ignorePatterns[i] = Pattern{regexp, str == pattern}
	}
	if !shouldIgnore(ignorePaths, ignorePatterns, "ignored file.iso") {
		t.Error("Should ignore this pattern")
	}
}


build-deb.sh
#!/bin/bash
set -euo pipefail

mntner="Lode Hoste <zillode@zillode.be>"
date=$(TZ=utc date +"%a, %d %b %Y %H:%M:%S %Z")
version=$(git describe --tags --always | sed 's/^v//')
ldflags="-w -X main.Version=$version"

build() { # os arch
	# Convert Debian format arch name to Go format
	arch="$2"
	if [[ $arch == "armhf" || $arch == "armel" ]]; then arch="arm"; fi
	if [[ $arch == "i386" ]]; then arch="386"; fi

	mkdir -p deb/usr/bin
	GOOS="$1" GOARCH="$arch" go build -v -i -ldflags "$ldflags" -o deb/usr/bin/syncthing-inotify
}

debianDir() { # arch
	arch="$1"

	rm -rf deb
	mkdir -p deb/DEBIAN

	cp debian/compat deb/DEBIAN/compat
	cat debian/control \
		| sed "s/{{arch}}/$arch/" \
		| sed "s/{{mntner}}/$mntner/" \
		| sed "s/{{version}}/$version/" \
		> deb/DEBIAN/control
	cat debian/changelog \
		| sed "s/{{date}}/$date/" \
		| sed "s/{{mntner}}/$mntner/" \
		| sed "s/{{version}}/$version/" \
		> deb/DEBIAN/changelog

	mkdir -p deb/lib/systemd
	mkdir -p deb/usr/lib/systemd
	cp -r etc/linux-systemd/system deb/lib/systemd
	cp -r etc/linux-systemd/user deb/usr/lib/systemd
}

# For each supported architecture (in Debian format), compile and pack into
# .deb archive.
for arch in amd64 i386 armhf armel ; do
	debianDir "$arch"
	build linux "$arch"
	dpkg-deb -b deb .
done


rlimit_unix.go
// Copyright (C) 2015 The Syncthing Authors.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// You can obtain one at http://mozilla.org/MPL/2.0/.

// +build !windows

package main

import "syscall"

// MaximizeOpenFileLimit tries to set the resoure limit RLIMIT_NOFILE (number
// of open file descriptors) to the max (hard limit), if the current (soft
// limit) is below the max. Returns the new (though possibly unchanged) limit,
// or an error if it was could not be changed.
func MaximizeOpenFileLimit() (int, error) {
	// Get the current limit on number of open files.
	var lim syscall.Rlimit
	if err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &lim); err != nil {
		return 0, err
	}

	// If we're already at max, there's no need to try to raise the limit.
	if lim.Cur >= lim.Max {
		return int(lim.Cur), nil
	}

	// Try to increase the limit to the max.
	oldLimit := lim.Cur
	lim.Cur = lim.Max
	if err := syscall.Setrlimit(syscall.RLIMIT_NOFILE, &lim); err != nil {
		return int(oldLimit), err
	}

	// If the set succeeded, perform a new get to see what happened. We might
	// have gotten a value lower than the one in lim.Max, if lim.Max was
	// something that indiciated "unlimited" (i.e. intmax).
	if err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &lim); err != nil {
		// We don't really know the correct value here since Getrlimit
		// mysteriously failed after working once... Shouldn't ever happen, I
		// think.
		return 0, err
	}

	return int(lim.Cur), nil
}


cross-compile.rb
#!/usr/bin/env ruby
# If you get the error "go build runtime: <runtime> must be bootstrapped using make.bash", please run:
# `/bin/bash -ic 'cd $(dirname $(dirname $(which go)))/src && GO386=387 GOARM=5 ./make.bash'`

oses = {
  "darwin" => ["amd64"],
  "dragonfly" => ["amd64"],
  "freebsd" => ["386", "amd64", "arm"],
  "linux" => ["386", "amd64", "arm"],
  "netbsd" => ["386", "amd64"],
  "openbsd" => ["386", "amd64"],
  "windows" => ["386", "amd64"]}

version = `git describe --abbrev=0 --tags`.chomp
diff = `git diff`.chomp

unless diff.empty?
  puts "Forgot to git reset --hard?"
  exit if gets.chomp != "n"
end

2.times do
  bootstrapped = false
  oses.each do |os, archs|
    next if ARGV.count > 0 && !ARGV.include?(os)
    archs.each do |arch|
      puts "building #{os}-#{arch}"
      name = "syncthing-inotify"
      newname = "syncthing-inotify-#{os}-#{arch}"
      cross = os.gsub(/-v\d/,"")
      if os.include?("windows")
        name = name + ".exe"
      end
      vars = "GOOS=#{os} GOARCH=#{arch}"
      if arch.include?("386")
        vars += " GO386=387"
      end
      if arch.include?("arm")
        vars += " GOARM=5"
      end
      ldflags = "-w -X main.Version=#{version}"
      build = "#{vars} go build -ldflags '#{ldflags}'"
      package = "tar -czf syncthing-inotify-#{os}-#{arch}-#{version}.tar.gz #{name}"
      remove = "rm #{name}"
      output = `#{build} 2>&1 && #{package} && #{remove}`
      puts output unless output.empty?
      if output.include?("must be bootstrapped") || output.include?("no such tool")
        `cd $(dirname $(which go))/../src && #{vars} ./make.bash --no-clean 1>&2`
        bootstrapped = true
      end
    end
  end
  exit unless bootstrapped
end
